app/experimenter/nimbus-ui/src/components/FormBranches/reducer.tsx

```diff
@@ -0,0 +1,266 @@
+import { useReducer, useMemo } from "react";
+import {
+  getExperiment_experimentBySlug,
+  getExperiment_experimentBySlug_treatmentBranches,
+} from "../../types/getExperiment";
+
+export function useFormBranchesReducer({
+  featureConfig,
+  referenceBranch,
+  treatmentBranches,
+}: getExperiment_experimentBySlug) {
+  const initialState = useMemo(() => {
+    let lastId = 0;
+
+    const equalRatio =
+      treatmentBranches !== null &&
+      treatmentBranches?.every(
+        (branch) => branch?.ratio === referenceBranch?.ratio,
+      );
+
+    const annotatedReferenceBranch =
+      referenceBranch === null
+        ? null
+        : annotateBranch("reference", referenceBranch!);
+
+    const annotatedTreatmentBranches = !Array.isArray(treatmentBranches)
+      ? null
+      : treatmentBranches.map((branch) => annotateBranch(lastId++, branch!));
+
+    return {
+      lastId,
+      equalRatio,
+      featureConfig,
+      referenceBranch: annotatedReferenceBranch,
+      treatmentBranches: annotatedTreatmentBranches,
+    };
+  }, [featureConfig, referenceBranch, treatmentBranches]);
+  return useReducer(formBranchesReducer, initialState);
+}
+
+export function formBranchesReducer(
+  state: FormBranchesState,
+  action: FormBranchesAction,
+): FormBranchesState {
+  switch (action.type) {
+    case "addBranch":
+      return addBranch(state);
+    case "removeBranch":
+      return removeBranch(state, action);
+    case "updateBranch":
+      return updateBranch(state, action);
+    case "removeFeatureConfig":
+      return removeFeatureConfig(state);
+    case "setFeatureConfig":
+      return setFeatureConfig(state, action);
+    case "setEqualRatio":
+      return setEqualRatio(state, action);
+    default:
+      return state;
+  }
+}
+
+export function extractSaveState(
+  state: FormBranchesState,
+): FormBranchesSaveState {
+  const { featureConfig, referenceBranch, treatmentBranches } = state;
+  return {
+    featureConfig,
+    referenceBranch: stripOneAnnotatedBranch(referenceBranch),
+    treatmentBranches: stripAnnotatedBranches(treatmentBranches),
+  };
+}
+
+export const REFERENCE_BRANCH_IDX = -1;
+
+export type FormBranchesState = Pick<
+  getExperiment_experimentBySlug,
+  "featureConfig"
+> & {
+  referenceBranch: null | AnnotatedBranch;
+  treatmentBranches: null | AnnotatedBranch[];
+  equalRatio: boolean;
+  lastId: number;
+};
+
+export type FormBranchesSaveState = Pick<
+  getExperiment_experimentBySlug,
+  "featureConfig" | "referenceBranch" | "treatmentBranches"
+>;
+
+export type FormBranchesAction =
+  | AddBranchAction
+  | RemoveBranchAction
+  | UpdateBranchAction
+  | RemoveFeatureConfigAction
+  | SetEqualRatioAction
+  | SetFeatureConfigAction;
+
+type AddBranchAction = {
+  type: "addBranch";
+};
+
+function addBranch(state: FormBranchesState) {
+  let { lastId, referenceBranch, treatmentBranches } = state;
+  lastId++;
```

Comment:
This is more of a nitpick and probably isn't worth the effort, but it would be cool if the count decremented when you removed a branch.

Reply:
It's not a count, it's an attempt at a unique ID. If the index goes down, React may re-use existing nodes in the DOM with the wrong data. That's the problem I was seeing when I brought in the ID code